File: codes/rust/112.路径总和.rs
/*
 * @lc app=leetcode.cn id=112 lang=rust
 *
 * [112] 路径总和
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        if root.is_none() {
            return false;
        }
        let root_val = root.as_ref().unwrap().borrow().val;
        let left = root.as_ref().unwrap().borrow().left.clone();
        let right = root.as_ref().unwrap().borrow().right.clone();
        if left.is_none() && right.is_none() && target_sum == root_val {
            return true;
        }
        Self::has_path_sum(left, target_sum - root_val)
            || Self::has_path_sum(right, target_sum - root_val)
    }
}
// @lc code=end

File: codes/python/112.路径总和.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
	def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
		# 如果根节点为空，直接返回False
		if not root:
			return False
		# 如果是叶子节点，判断是否等于目标值
		if root.left is None and root.right is None:
			return root.val == targetSum
		# 递归判断左子树和右子树
		left_has_path_sum = self.hasPathSum(root.left, targetSum-root.val)
		right_has_path_sum = self.hasPathSum(root.right, targetSum-root.val)
		return left_has_path_sum or right_has_path_sum